src/main/java/net/minestom/server/lock/Acquirable.java

```diff
@@ -0,0 +1,121 @@
+package net.minestom.server.lock;
+
+import net.minestom.server.entity.Entity;
+import net.minestom.server.thread.BatchThread;
+import net.minestom.server.thread.ThreadProvider;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Objects;
+import java.util.function.Consumer;
+
+/**
+ * Represents an element which can be acquired.
+ * Used for synchronization purpose.
+ *
+ * @param <T> the acquirable object type
+ */
+public final class Acquirable<T> {
+
+    public static final ThreadLocal<Collection<Entity>> CURRENT_ENTITIES = ThreadLocal.withInitial(Collections::emptyList);
+
+    public static @NotNull Collection<@NotNull Entity> currentEntities() {
+        return CURRENT_ENTITIES.get();
+    }
+
+    @ApiStatus.Internal
+    public static void refreshEntities(@NotNull Collection<@NotNull Entity> entities) {
+        CURRENT_ENTITIES.set(entities);
+    }
+
+    private final T value;
+    private final Handler handler;
+
+    public Acquirable(@NotNull T value) {
+        this.value = value;
+        this.handler = new Handler();
+    }
+
+    /**
+     * Blocks the current thread until 'this' can be acquired,
+     * and execute {@code consumer} as a callback with the acquired object.
+     *
+     * @param consumer the acquisition consumer
+     */
+    public void acquire(@NotNull Consumer<@NotNull T> consumer) {
+        final Thread currentThread = Thread.currentThread();
+        final BatchThread elementThread = getHandler().getBatchThread();
+        Acquisition.acquire(currentThread, elementThread, () -> consumer.accept(unwrap()));
+    }
+
+    /**
+     * Executes {@code consumer} only if this element can be safely
+     * acquired without any synchronization.
+     *
+     * @param consumer the acquisition consumer
+     * @return true if the acquisition happened without synchronization,
+     * false otherwise
+     */
+    public boolean tryAcquire(@NotNull Consumer<@NotNull T> consumer) {
```

Comment:
This could have a method that doesn't require a consumer that returns the acquired element or null.